xds/internal/server/rds_handler.go

```diff
@@ -0,0 +1,137 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package server
+
+import (
+	"sync"
+
+	"google.golang.org/grpc/xds/internal/xdsclient"
+)
+
+// rdsHandlerUpdate wraps the full RouteConfigUpdate that are dynamically queried for a given
+// server side listener.
+type rdsHandlerUpdate struct {
+	updates map[string]xdsclient.RouteConfigUpdate
+	err     error
+}
+
+// rdsHandler handles any RDS queries that need to be started for a given server
+// side listeners Filter Chains (i.e. not inline).
+type rdsHandler struct {
+	parent *listenerWrapper
```

Comment:
Why do we need to pass the `listenerWrapper` here? All we need is a function to register an RDS watch. It is in general good practice to only provide as much access as required.

Similarly, it feels weird to have `updateChannel` created in `newRDSHandler` and then `listenerWrapper` directly accesses that field. Again, this can be avoided by having the `updateChannel` injected into this type in its ctor.

```golang
type rdsHandler struct {
	watchRouteConfig func(xdsclient.RouteConfigUpdate, error)
	updateChannel chan rdsHandlerUpdate

	mu sync.Mutex
	rdsUpdates map[string]xdsclient.RouteConfigUpdate
	rdsCancels map[string]func()
}
```

```golang
func newRDSHandler(watch func(xdsclient.RouteConfigUpdate, error), ch chan rdsHandlerUpdate) *rdsHandler {
	return &rdsHandler{
		watchRouteConfig: watch,
		updateChannel: ch,
		rdsUpdates:    make(map[string]xdsclient.RouteConfigUpdate),
		rdsCancels:    make(map[string]func()),
	}
}
```

Reply:
I find holding onto a function watchRouteConfig weird. If anything, you should pass it an xdsClient reference/pointer, rather than the whole ListenerWrapper. I patterned this after how cluster handler is linked with cds balancer. But I don't mind holding a reference to the parent for the parents xdsclient. For the second one, I personally still prefer the way I did it (again, how me/Menghan did in in cds balancer). I like the idea that a component holds onto a channel it itself creates, and we receive from that channel, which to me seems like it logically ties that channel to the component. However, if you feel strongly about the second one let me know. I can also switch the update channel to a buffer.Unbounded, then on a recv, typecast it to determine update.

Reply:
I understand the `cdsBalancer` does things the way it is done here. I actually looked at that code as well as part of reviewing this, and I felt it should not have been done that way.

This is also another one of these design patterns where the dependencies of a component are injected at creation time instead of the component creating those dependencies by itself. This decouples the components as much as possible, and eases the testing of each component and in general reduces maintenance burden over time as the two components can evolve independently (as they are lightly coupled).

I would still prefer a vanilla channel over a `buffer.Unbounded` since we are only interested in the most recent update.

If you still feel that the channel should be owned/created by the `rdsHandler`, let's discuss.

Reply:
xdsClient makes more sense, as needed multiple watches to start. I like passing a reference to component xdsClient too I guess, rather than doing A.B. I'll switch it to that. Ok, I get the point about decoupling, as right now listener wrapper has to do rdsHandler.field rather than it's own field. In regards to vanilla channel, what do you mean by that? I plan on passing rds handler the same channel that lds update is written to, and on the event of a receive, typecast it. Or do you think we should have a separate channel for lds and rds?

Reply:
Sure `xdsClient` is fine.

By vanilla channel, I meant something of type `chan <foo>` and not something that is a wrapper on top of a channel like a `buffer.Unbounded`.

If you pass the same channel as the one which is used for lds updates, will be you able to do the drain and push operation on it (without possibly erasing an update of the other type)?